seonoh/ArchitectureStudy/app/src/main/java/com/example/seonoh/seonohapp/MainActivity.kt

```diff
@@ -1,12 +1,108 @@
+import androidx.appcompat.app.AppCompatActivity
+import com.example.seonoh.seonohapp.model.Market
+import com.example.seonoh.seonohapp.network.CoinRequest
+import com.example.seonoh.seonohapp.network.RetrofitCreator
+import com.google.android.material.tabs.TabLayout
+import kotlinx.android.synthetic.main.activity_main.*
 
-class MainActivity : AppCompatActivity() {
+class MainActivity : AppCompatActivity(), CoinRequest.BaseResult<ArrayList<Market>> {
+
+    private lateinit var mPagerAdapter: TabPagerAdapter
+    private val TAG = "COIN_MAIN"
+    private lateinit var toast: Toast
+    private lateinit var conMarketNameList: List<String>
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        toast = Toast.makeText(this, "뒤로가기를 한번 더 누르시면 앱이 종료됩니다.", Toast.LENGTH_SHORT)
+        CoinRequest(RetrofitCreator.coinApi).marketSend(this@MainActivity)
+    }
+
+    fun initView(data: ArrayList<String>) {
+        mPagerAdapter = TabPagerAdapter(supportFragmentManager).apply {
+            setData(data)
+            notifyDataSetChanged()
+        }
+
+        coinViewPager.adapter = mPagerAdapter
+
+        coinViewPager.addOnPageChangeListener(object : TabLayout.TabLayoutOnPageChangeListener(tabLayout) {
+            override fun onPageSelected(position: Int) {
+                super.onPageSelected(position)
+
+                when (position) {
+                    0 -> {
+                        Log.e(TAG, "KRW")
+                    }
+
+                    1 -> {
+                        Log.e(TAG, "BTC")
+
+                    }
+
+                    2 -> {
+                        Log.e(TAG, "ETH")
+
+                    }
+                    3 -> {
+                        Log.e(TAG, "USDT")
+
+                    }
+                }
+            }
+        })
+
+        tabLayout.apply {
+
+            conMarketNameList.forEach {
+                addTab(tabLayout.newTab().setText(it))
+            }
+
+            addOnTabSelectedListener(TabLayout.ViewPagerOnTabSelectedListener(coinViewPager))
+        }
+    }
+
+    override fun onBackPressed() {
+        if (toast.view.isShown) {
+            finish()
+        } else {
+            toast.show()
+        }
     }
+
+    private fun classifyMarketData(marketData: ArrayList<Market>): ArrayList<String> {
+
+        conMarketNameList = marketData.map {
+            it.market.substringBefore("-")
+        }.distinct()
+
+        val marketDataList = ArrayList<String>()
+
+        conMarketNameList.forEach { title ->
+
+            marketDataList += (marketData.filter {
+                it.market.substringBefore("-") == title
+            }.joinToString(",") {
+                it.market
+            })
+        }
+
+        return marketDataList
+    }
+
+    override fun getNetworkSuccess(result: ArrayList<Market>) {
+        val data = classifyMarketData(result)
+        initView(data)
```

Comment:
api 응답으로 받은 데이터만 넣어주면 되기 때문에 setData 만 해주는게 좋을 것 같습니다
```suggestion
        mPagerAdapter.setData(data)
```

Reply:
네 감사합니다 반영했습니다!

Reply:
@seonoh 
반영 커밋 남겨주세요

Reply:
d3de59a